lib/Factory/LayoutFactory.php

```diff
@@ -1397,6 +1411,29 @@ public function createFromZip(
         $widgets = $layout->getAllWidgets();
         $this->getLog()->debug('Layout has ' . count($widgets) . ' widgets');
 
+        // Go through all the widgets and upgrade from v3 to v4
+        foreach ($widgets as $widget) {
+            // Load the widget
+            $widget->load();
+
+            $module = $this->moduleFactory->getByType($widget->type);
+            if ($module->isWidgetCompatibilityAvailable()) {
+                // Grab a widget compatibility interface, if there is one
+                $widgetCompatibilityInterface = $module->getWidgetCompatibilityOrNull();
+                if ($widgetCompatibilityInterface !== null) {
+                    try {
+                        $upgraded = $widgetCompatibilityInterface->upgradeWidget($widget, $widget->schemaVersion, 2);
+                        if ($upgraded) {
+                            $widget->schemaVersion = 2;
+                            $widget->save(['alwaysUpdate'=>true]);
```

Comment:
It might be wise to explicitly say `notifyDisplays => false` here so that its clear we are not expecting an upgraded widget to trigger a display update?